crates/client/eth/src/client.rs

```diff
@@ -0,0 +1,172 @@
+use bitvec::macros::internal::funty::Fundamental;
+use dc_db::DeoxysBackend;
+use std::sync::Arc;
+
+use dc_sync::metrics::block_metrics::BlockMetrics;
+use dp_convert::{ToFelt, ToStarkFelt};
+use dp_transactions::TEST_CHAIN_ID;
+use dp_utils::channel_wait_or_graceful_shutdown;
+use futures::StreamExt;
+use primitive_types::H256;
+use serde_json::Value;
+use starknet_api::hash::StarkFelt;
+use starknet_types_core::felt::Felt;
+use url::Url;
+
+pub struct EthereumClient {
+    provider: Arc<ReqwestProvider>,
+    l1_core_address: Address,
+}
+
+sol!(
+    #[allow(missing_docs)]
+    #[sol(rpc)]
+    StarknetCore,
+    "src/abis/starknet_core.json"
+);
+
+impl EthereumClient {
+    /// Create a new EthereumClient instance with the given RPC URL
+    pub async fn new(url: Url, l1_core_address: Address) -> anyhow::Result<Self> {
+        let provider = ProviderBuilder::new().on_http(url);
+
+        Ok(Self { provider: Arc::new(provider), l1_core_address })
+    }
+
+    /// Retrieves the latest Ethereum block number
+    pub async fn get_latest_block_number(&self) -> anyhow::Result<u64> {
+        let block_number = self.provider.get_block_number().await?.as_u64();
+        Ok(block_number)
+    }
+
+    /// Get the block number of the last occurrence of a given event.
+    pub async fn get_last_event_block_number(&self) -> anyhow::Result<u64> {
+        let topic = B256::from_slice(&hex::decode(&LOG_STATE_UPDTATE_TOPIC[2..])?);
+        let latest_block: u64 =
+            self.get_latest_block_number().await.expect("Failed to retrieve latest block number").into();
+
+        // Assuming an avg Block time of 15sec we check for a LogStateUpdate occurence in the last ~24h
+        let filter = Filter::new()
+            .from_block(latest_block - 6000)
+            .to_block(latest_block)
+            .address(self.l1_core_address)
+            .event_signature(topic);
+
+        let logs = self.provider.get_logs(&filter).await?;
+
+        if let Some(last_log) = logs.last() {
+            let last_block: u64 = last_log.block_number.context("no block number in log")?;
+            Ok(last_block)
+        } else {
+            bail!("no event found")
+        }
+    }
+
+    /// Get the last Starknet block number verified on L1
+    pub async fn get_last_block_number(&self) -> anyhow::Result<u64> {
+        let client = self.provider.clone();
+        let address: Address = self.l1_core_address;
+        let contract = StarknetCore::new(address, client);
+
+        let block_number = contract.stateBlockNumber().call().await.unwrap();
+        let last_block_number: u64 = (block_number._0).as_u64();
+        Ok(last_block_number)
+    }
+
+    /// Get the last Starknet state root verified on L1
+    pub async fn get_last_state_root(&self) -> anyhow::Result<StarkFelt> {
+        let client = self.provider.clone();
+        let address: Address = self.l1_core_address;
+        let contract = StarknetCore::new(address, client);
```

Comment:
should we just store `StarknetCore` instead of the address inside `EthereumClient`

Reply:
good idea, we won't have to create one everytime